core/ledger/ledger.go

```diff
@@ -179,68 +178,81 @@ func (self *Ledger) AddHeaders(headers []*types.Header) error {
 func (self *Ledger) AddBlock(block *types.Block, stateMerkleRoot common.Uint256) error {
 	if len(DefLedgerMgr.Ledgers) > 1 && self.ShardID.IsRootShard() {
 		return self.ChildLedger.ParentBlockCache.PutBlock(block, stateMerkleRoot)
-	} else {
-		err := self.ldgStore.AddBlock(block, stateMerkleRoot)
-		if err != nil {
-			log.Errorf("Ledger AddBlock BlockHeight:%d BlockHash:%x error:%s", block.Header.Height, block.Hash(), err)
+	}
+
+	if self.ParentLedger != nil {
+		currentParentHeight := self.ParentLedger.GetCurrentBlockHeight()
+		if block.Header.ParentHeight > currentParentHeight {
+			// check if parent block available
+			for h := currentParentHeight + 1; h <= block.Header.ParentHeight; h++ {
+				if _, _, err := self.ParentBlockCache.GetBlock(h); err != nil {
+					return fmt.Errorf("shard %d get parent block %d failed: %s", self.ShardID, h, err)
+				}
+			}
 		}
-		return err
 	}
+
+	// FIXME:
+	// 1. ExecuteBlock/SubmitBlock requires saving block
+	// 2. lock released and re-acquired after ExecuteBlock(block)
+	execResult, err := self.ExecuteBlock(block)
+	if err != nil {
+		return fmt.Errorf("shard %d execute block %d failed: %s", self.ShardID, block.Header.Height, err)
+	}
+
+	if err := self.SubmitBlock(block, execResult); err != nil {
+		return fmt.Errorf("shard %d submit block %d failed: %s", self.ShardID, block.Header.Height, err)
+	}
+
+	return nil
 }
 
 func (self *Ledger) ExecuteBlock(b *types.Block) (store.ExecuteResult, error) {
 	if !self.ShardID.IsRootShard() {
-		parentBlock, merkleRoot, err := self.ParentBlockCache.GetBlock(b.Header.ParentHeight)
-		if err != nil {
-			if err == scommon.ErrNotFound {
-				return self.ldgStore.ExecuteBlock(b)
-			} else {
-				log.Errorf("Ledger ExecuteBlock GetBlock sharad height:%d,ParentHeight:%d error:%s", b.Header.Height, b.Header.ParentHeight, err)
+		currentParentHeight := self.ParentLedger.GetCurrentBlockHeight()
+		for ph := currentParentHeight + 1; ph <= b.Header.ParentHeight; ph++ {
+			parentBlock, merkleRoot, err := self.ParentBlockCache.GetBlock(ph)
+			if err != nil {
+				log.Errorf("Ledger %d ExecuteBlock GetBlock shard height:%d,ParentHeight:%d error:%s", self.ShardID, b.Header.Height, ph, err)
 				return store.ExecuteResult{}, err
 			}
+			result, err := self.ParentLedger.ldgStore.ExecuteBlock(parentBlock)
```

Comment:
this will have problem, previous block result is not submitted, so will cause executing block at height `H+2` based on ledger state at `H`

Reply:
right.